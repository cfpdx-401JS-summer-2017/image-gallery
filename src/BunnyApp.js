import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { plusBunny, minusBunny, fetchBunnies} from './services/bunnies';
import {Thumbnail, List, Gallery} from './components/viewFormats';
import {AddBunny, eliminateBunny, DeleteBunny} from './components/addRemoveBunny';
import qs from 'qs';

function genBunnyList (viewtype, Component) {
    return function bunnyView({ bunnies, onDelete }) {
        bunnyView.PropTypes = {
            bunnies:PropTypes.array.isRequired,
            onRemove:PropTypes.func
        };
        return <div> 
            <div style={{fontWeight:'bold'}}>Bunny {viewtype} </div> 
            <ul>
                {bunnies && bunnies.map(bunny => (
                    <li key={bunny._id}>
                        <Component bunny={bunny} /> 
                        <DeleteBunny onDelete={() => onDelete(bunny)} bunnies={bunnies}/>
                    </li>
                ))}
            </ul>
        </div>;
    };
}
genBunnyList.PropTypes = {
    viewtype:PropTypes.string.isRequired,
    Component:PropTypes.object.isRequired
};
const bunnyDetail = genBunnyList('Detail',List);
const bunnyThumbnail = genBunnyList('Thumbnail',Thumbnail);
function bunnyGallery({ bunny, onDelete }) {
    return <div>
        <div style={{fontWeight:'bold'}}>Bunny Gallery </div>
        <ul>
        
            <Gallery bunny={bunny}/>
            <DeleteBunny onDelete={() => onDelete(bunny)}/>
        </ul>
    </div>;
}
const View = {
    detail: bunnyDetail,
    thumbnail: bunnyThumbnail,
    gallery: bunnyGallery
};
const views = Object.keys(View);
class BunnyApp extends Component {
    constructor() {
        super();
        this.state = {
            bunnies: [],
            view: views[1],
            views: views,
            i: 0,
            bunnyNum: 0,
            loading: true
            
        };
    }
    removeBunny = (bunny) => {
        minusBunny(bunny._id)
            .then(() => {
                this.setState(prevState => ({ 
                    bunnies:eliminateBunny(prevState.bunnies, bunny)
                }));

            });
    }
    addBunny = (title,description,url) => {
        plusBunny(title, description, url)
            .then((res) => {
                this.setState(prevState =>({
                    bunnies:[...prevState.bunnies,res.body]
                }));
            });
    }
    handleChange(target) {
        this.setState({
            title: target.value
        });
    }
    componentDidMount() {
        fetchBunnies()
            .then(res => {
                this.setState(() => ({
                    bunnies: res,
                    loading: false
                }));
            })            
            .catch(error => console.log(error));
    }
    render() {
        const { loading } = this.state;
        if(loading) return <div>Loading...</div>;
        const { bunnies } =this.state;
        const { views } = this.state;
        const queried = qs.parse(this.props.location.search.slice(1));
        const BunnyView = View[queried.view||'thumbnail'];
        let { i } = this.state;
        const bunny = bunnies[i];
    
        return (
            <main>
                <header>
                    <h1>Bunny App</h1>
                </header>
                <section>
                    {views.map(v => (
                        <button key={v} 
                            onClick={() => {this.props.history.push(`/photos?view=${v}`);}}>
                            {v}
                        </button>
                    ))}
                    {queried.view === 'gallery' && <section>
                        <button
                            onClick={() => this.setState({i:this.state.i-1})}
                            style={{padding: '10px'}}
                            disabled={this.state.i === 0}
                        >previous
                        </button>
                        <button 
                            onClick={() =>  this.setState({i:this.state.i+1})}
                            style={{padding: '10px'}}
                            disabled={this.state.i===bunnies.length-1}
                        >next
                        </button>
                    </section>
                    }
                    <BunnyView bunnies={bunnies} bunny={bunny} onDelete={this.removeBunny}/>                    
                </section>
                <section>
                    <AddBunny onAdd={this.addBunny}/>
                </section>
            </main>
        );
    }
}
export default BunnyApp;
