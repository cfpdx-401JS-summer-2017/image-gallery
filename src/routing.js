import React from 'react';
import {BrowserRouter as Router, Route, Link} from 'react-router-dom';
import BunnyApp from './BunnyApp';

const Home = () => (
    <div>
        <h2>Home</h2>
    </div>
);

const About = () => (
    <div>
        <h2>About</h2>
    </div>
);
export const TopBar = () => (
    
    <div>
        <ul>
            <li><Link to="/">Home</Link></li>
            <li><Link to="/about">About</Link></li>
            <li><Link to="/photos">photos</Link></li>
        </ul>

        <hr/>

        <Route exact path="/" component={Home}/>
        <Route exact path="/about" component={About}/>
        <Route path="/photos" component={BunnyApp} />
    </div>
);

export default () => <Router><TopBar/></Router>;