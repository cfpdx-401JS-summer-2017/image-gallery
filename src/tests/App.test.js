import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import BunnyApp from '../BunnyApp';
import { BunnyView } from '../BunnyApp';
import { TopBar } from'../routing';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
it('renders a snapshot', () => {
    const tree = renderer.create(<BunnyApp/>).toJSON();
    expect(tree).toMatchSnapshot();
});
describe('Routes', () => {
    it('Defaults to home', () => {
        const wrapper = mount( 
            <MemoryRouter>
                <TopBar/>
            </MemoryRouter>
        );
        const home = wrapper.find('Home');
        expect(home.length).toBe(1);
    });
    it('/about', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/about']} >
                <TopBar/>
            </MemoryRouter>
        );
        const about = wrapper.find('About');
        expect(about.length).toBe(1);
    });
});

it('renders without crashing', () => {
  
    const div = document.createElement('div');
    ReactDOM.render(<BunnyApp />, div);
});

describe('BunnyApp', () => {
    function testView(view) {
        it(`shows ${view}`, () => {
            const wrapper = shallow(<BunnyApp view={view} bunnies={[]} />);
            expect(toJSON(wrapper)).toMatchSnapshot();
        });
    }
    testView('gallery');
    testView('detail');
    testView('thumbnail');
});


