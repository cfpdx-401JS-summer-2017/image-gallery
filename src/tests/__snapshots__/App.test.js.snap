// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BunnyApp shows detail 1`] = `
<div>
  Loading...
</div>
`;

exports[`BunnyApp shows gallery 1`] = `
<div>
  Loading...
</div>
`;

exports[`BunnyApp shows thumbnail 1`] = `
<div>
  Loading...
</div>
`;

exports[`renders a snapshot 1`] = `
<div>
  Loading...
</div>
`;
